### PR TITLE
feat: WalletConnect v2 support

### PR DESCRIPTION
**Pull request by jxom**
Co-authored-by: Gancho Radkov <ganchoradkov93@gmail.com>

Source PR: https://github.com/wagmi-dev/wagmi/pull/1208

## Description

Add WalletConnect v2 support to the existing WC Connector.

## Additional Information
WalletConnect v1 is to be depreciated on March 1, 2023.
This v2 connector will allow wagmi users to have enough time to prepare and migrate to WalletConnect v2 which offers significantly better developer & user experience

- [x] I read the [contributing docs](/wagmi-dev/wagmi/blob/main/.github/CONTRIBUTING.md) (if this is your first contribution)

Your ENS/address:

### CODE REVIEW

File: packages/connectors/src/walletConnect.ts [hunk 1]
```diff
@@ -1,47 +1,62 @@
...
+  qrcode?: boolean
+} & (
+  | ({
+      version?: '1'
+    } & ConstructorParameters<typeof WalletConnectProvider>[0])
+  | ({
+      /**
+       * Project ID associated with [WalletConnect account](https://cloud.walletconnect.com)
+       */
+      projectId: NonNullable<UniversalProviderOpts['projectId']>
```
Comment (seendsouza) [inline packages/connectors/src/walletConnect.ts:33]: Isn't it possible to have no projectId if using your own relayer for walletconnect v2? https://github.com/WalletConnect/walletconnect-monorepo/blob/55ac00aa96a322c8a229d841ee262bab81818c91/providers/universal-provider/src/types/misc.ts#L9

Comment (0xAsimetriq) [reply packages/connectors/src/walletConnect.ts:33]: Just a heads up, relayer server for v2 is still work in progress, so for a bit users will still need to use managed solution.

File: package.json [hunk 2]
```diff
@@ -33,6 +33,7 @@
     "execa": "^6.1.0",
     "fs-extra": "^11.1.0",
     "lint-staged": "^13.0.3",
+    "lokijs": "^1.5.12",
```
Comment (tmm) [inline package.json:36]: What is this for?

File: packages/connectors/src/walletConnect.ts [hunk 3]
```diff
@@ -81,35 +197,66 @@ export class WalletConnectConnector extends Connector<
...
   }
 
   async disconnect() {
     const provider = await this.getProvider()
-    await provider.disconnect()
+    try {
+      await provider.disconnect()
+    } catch (error) {
+      // If the session does not exist, we don't want to throw.
+      if (/No matching key/i.test((error as Error).message)) return
```
Comment (tmm) [inline packages/connectors/src/walletConnect.ts:221]: Probably don't want to `return` here so we can remove the event listeners later on, right?

File: packages/connectors/src/walletConnect.ts [hunk 4]
```diff
@@ -81,35 +197,66 @@ export class WalletConnectConnector extends Connector<
...
   }
 
   async getChainId() {
     const provider = await this.getProvider()
-    const chainId = normalizeChainId(provider.chainId)
-    return chainId
+    if (this.version === '1')
+      return normalizeChainId((provider as WalletConnectProvider).chainId)
+    return (
+      getClient().data?.chain?.id ??
```
Comment (tmm) [inline packages/connectors/src/walletConnect.ts:257]: Why do we want to prefer the client's chain id ahead of what the connector returns? Can we do the opposite (request chain id from the provider) and fallback to the client?

**Review by tmm — COMMENTED** [review]
Going to play around with a bit today and see if I can clean up the version handling internally with generic.

File: packages/connectors/src/walletConnect.ts (hunk 3 above)
Comment (jxom) [reply packages/connectors/src/walletConnect.ts:221]: Good catch!

File: package.json (hunk 2 above)
Comment (jxom) [reply package.json:36]: Build was [failing](https://github.com/wagmi-dev/references/actions/runs/3709955301/jobs/6319735218#step:6:20).

**Pull request merged**